source/glimmer-components/adding-behavior-and-keeping-states.md

```diff
@@ -0,0 +1,222 @@
+Most HTML elements fufills one or more of these purposes: describing the
+semantic significance of its content, altering their presentation and adding
+behavior.
+
+Take the `<article>` tag for instance: it has no default style rules, but it
+provides a semantic description its content. The `<b>` tag on the other hand,
+changes the presentation of its content (bolding the text) without conveying
+any semantic meanings. Of course, there are also many tags that serves both of
+these purposes at the same time, such as the `<ol>` and `<li>` pair. All the
+Components we have written so far fall into this category.
+
+Besides semantic and presentational elements, there are also HTML elements
+that add *behavior* for its content. For example, clicking on a `<label>` tag
+would move the focus to the corresponding input field; likewise, clicking on an
+`<a>` tag would navigate the browser to the URL specified in the `href`
+attribute.
+
+Ember allows us to add behavior to our Components using JavaScript. In
+additional to a Handlebars template, each Component can also have a
+corresponding JavaScript file in the `app/components/` directory.
+
+As an example, let's implement a `<random-color>` Component gives its content
+a solid fill or a "glow" effect in a different color everytime you click on it:
+
+```app/templates/application.hbs
+<random-color fill={{true}}>So</random-color>
+<random-color glow={{true}}>many</random-color>
+<random-color fill={{true}} glow={{true}}>colors</random-color>!!!
+```
+
+```app/templates/components/random-color.hbs
+<span>{{yield}}</span>
+```
+
+```app/components/random-color.js
+import Ember from 'ember';
+
+// Returns a random integer from 0 up to (and excluding) max
+function randomInteger(max) {
+  return Math.floor( Math.random() * max );
+}
+
+// Returns a random color in rgb(x,y,z) format
+function randomColor() {
+  let r = randomInteger(256),
+      g = randomInteger(256),
+      b = randomInteger(256);
+
+  return `rgb(${r},${g},${b})`;
+}
+
+// Returns a text-shadow for the "glow" effect
+function randomGlow() {
+  return `0px 0px 4px ${ randomColor() }`;
+}
+
+export default Ember.GlimmerComponent.extend({
+
+  setColor() {
+    if (this.attrs.fill) {
+      this.element.style.color = randomColor();
+    }
+
+    if (this.attrs.glow) {
+      this.element.style.textShadow = randomGlow();
+    }
+  },
+
+  didInsertElement() {
+    this.setColor();
+  },
+
+  click() {
+    this.setColor();
+  }
+
+});
+```
+
+You can see the result [at this Ember twiddle](http://ember-twiddle.com/e55043ce86b0a7fd2117).
```

Comment:
> Mixed Content: The page at 'about:srcdoc' was loaded over HTTPS, but requested an insecure script 'http://builds.emberjs.com/canary/ember.debug.js'. This request has been blocked; the content must be served over HTTPS.
VM385 about:srcdoc:1 Mixed Content: The page at 'about:srcdoc' was loaded over HTTPS, but requested an insecure script 'http://builds.emberjs.com/canary/ember-template-compiler.js'. This request has been blocked; the content must be served over HTTPS.